### PR TITLE
as sugested by @zz123happy make cache_index_and_filter_blocks configurable

### PR DESCRIPTION
**Pull request by dothebart**
As sugested in https://github.com/arangodb/arangodb/issues/11577 add a configuration flag. 

https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager upstream flag documentation.

### CODE REVIEW

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 1]
```diff
@@ -335,6 +336,17 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
...
                      new BooleanParameter(&_enforceBlockCacheSizeLimit));
 
+  options->addOption(
+      "--rocksdb.cache-index-and-filter-blocks",
+      "In most cases, actually used blocks in block cache are just a small "
+      "percentage than data cached in block cache, so when users enable this "
+      "feature, the block cache capacity will cover the memory usage for both "
+      "of block cache and memtable. "
+      "If users also enable cache_index_and_filter_blocks, then the three major "
+      "uses of memory of RocksDB will be capped by the single cap",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:346]:
```suggestion
  options->addOption(
      "--rocksdb.cache-index-and-filter-blocks",
      "if turned on, the RocksDB block cache quota will also include RocksDB memtable sizes",
```

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 2]
```diff
@@ -553,6 +553,10 @@ void RocksDBEngine::start() {
     tableOptions.no_block_cache = true;
   }
   tableOptions.cache_index_and_filter_blocks = opts._cacheIndexAndFilterBlocks ;
+  tableOptions.cache_index_and_filter_blocks_with_high_priority = opts._cacheIndexNFilterBlocksWithHighPriority;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:556]:
```suggestion
  tableOptions.cache_index_and_filter_blocks_with_high_priority = opts._cacheIndexAndFilterBlocksWithHighPriority;
```

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 3]
```diff
@@ -553,6 +553,10 @@ void RocksDBEngine::start() {
...
+  tableOptions.pin_l0_filter_and_index_blocks_in_cache = opts._pinTopLevelIndexAndFilter;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:557]:
```suggestion
  tableOptions.pin_l0_filter_and_index_blocks_in_cache = opts._pinl0FilterAndIndexBlocksInCache;
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 4]
```diff
@@ -116,7 +116,11 @@ RocksDBOptionFeature::RocksDBOptionFeature(application_features::ApplicationServ
       _level0StopTrigger(rocksDBDefaults.level0_stop_writes_trigger),
       _recycleLogFileNum(rocksDBDefaults.recycle_log_file_num),
       _enforceBlockCacheSizeLimit(false),
-      _cacheIndexAndFilterBlocks(false),
+      _cacheIndexAndFilterBlocks(rocksDBTableOptionsDefaults.cache_index_and_filter_blocks),
+      _cacheIndexNFilterBlocksWithHighPriority(
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:120]:
```suggestion
      _cacheIndexAndFilterBlocksWithHighPriority(
```

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10172/

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 5]
```diff
@@ -116,7 +116,11 @@ RocksDBOptionFeature::RocksDBOptionFeature(application_features::ApplicationServ
...
+        rocksDBTableOptionsDefaults.cache_index_and_filter_blocks_with_high_priority),
+      _pinl0FilterNIndexBlocksInCache(rocksDBTableOptionsDefaults.pin_l0_filter_and_index_blocks_in_cache),
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:122]:
```suggestion
      _pinl0FilterAndIndexBlocksInCache(rocksDBTableOptionsDefaults.pin_l0_filter_and_index_blocks_in_cache),
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 6]
```diff
@@ -342,6 +346,32 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
       new BooleanParameter(&_cacheIndexAndFilterBlocks),
       arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden));
 
+  options->addOption(
+      "--rocksdb.cache-index-and-filter-blocks-with-high-priority",
+      "If cache_index_and_filter_blocks is enabled, cache index and filter blocks with high priority. "
+      "If set to true, depending on implementation of block cache, "
+      "index and filter blocks may be less likely to be evicted than data blocks.",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:353]:
```suggestion
      "if true and `--rocksdb.cache-index-and-filter-blocks` is also true, cache index and filter blocks with high priority, "
      "making index and filter blocks be less likely to be evicted than data blocks",
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 7]
```diff
@@ -342,6 +346,32 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
...
+      new BooleanParameter(&_cacheIndexNFilterBlocksWithHighPriority),
+      arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden));
+
+  options->addOption(
+      "--rocksdb.pin-l0-filter-and-index-blocks-in-cache",
+      "if cache_index_and_filter_blocks is true and the below is true, "
+      "then filter and index blocks are stored in the cache, "
+      "but a reference is held in the 'table reader' object "
+      "so the blocks are pinned and only evicted from cache when the table reader is freed.",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:362]:
```suggestion
      "if true and `--rocksdb.cache-index-and-filter-blocks is also true, "
      "filter and index blocks are pinned and only evicted from cache when the table reader is freed",
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 8]
```diff
@@ -342,6 +346,32 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
...
+      new BooleanParameter(&_pinTopLevelIndexAndFilter),
+      arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden));
+
+  options->addOption(
+      "--rocksdb.pin-top-level-index-and-filter",
+      "If cache_index_and_filter_blocks is true and the below is true, "
+      "then the top-level index of partitioned filter and index blocks are stored in the cache, "
+      "but a reference is held in the 'table reader' object so the blocks are pinned "
+      "and only evicted from cache when the table reader is freed. This is not limited to l0 in LSM tree.",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:371]:
```suggestion
      "If true and `--rocksdb.cache-index-and-filter-blocks` is also true, "
      "the top-level index of partitioned filter and index blocks are pinned and only evicted from cache when the table reader is freed",
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 9]
```diff
@@ -342,6 +346,32 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
       new BooleanParameter(&_cacheIndexAndFilterBlocks),
       arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden));
 
+  options->addOption(
+      "--rocksdb.cache-index-and-filter-blocks-with-high-priority",
+      "If cache_index_and_filter_blocks is enabled, cache index and filter blocks with high priority. "
+      "If set to true, depending on implementation of block cache, "
+      "index and filter blocks may be less likely to be evicted than data blocks.",
+      new BooleanParameter(&_cacheIndexNFilterBlocksWithHighPriority),
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:354]:
```suggestion
      new BooleanParameter(&_cacheIndexAndFilterBlocksWithHighPriority),
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 10]
```diff
@@ -342,6 +346,32 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
...
+      arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden));
+
+  options->addOption(
+      "--rocksdb.pin-l0-filter-and-index-blocks-in-cache",
+      "if cache_index_and_filter_blocks is true and the below is true, "
+      "then filter and index blocks are stored in the cache, "
+      "but a reference is held in the 'table reader' object "
+      "so the blocks are pinned and only evicted from cache when the table reader is freed.",
+      new BooleanParameter(&_pinTopLevelIndexAndFilter),
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:363]:
```suggestion
      new BooleanParameter(&_pinl0FilterAndIndexBlocksInCache),
```

File: arangod/StorageEngine/RocksDBOptionFeature.h [hunk 11]
```diff
@@ -83,6 +83,9 @@ class RocksDBOptionFeature final : public application_features::ApplicationFeatu
   bool _recycleLogFileNum;
   bool _enforceBlockCacheSizeLimit;
   bool _cacheIndexAndFilterBlocks;
+  bool _cacheIndexNFilterBlocksWithHighPriority;
+  bool _pinl0FilterNIndexBlocksInCache;
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.h:87]:
```suggestion
  bool _cacheIndexAndFilterBlocksWithHighPriority;
  bool _pinl0FilterAndIndexBlocksInCache;
```

**Comment by jsteemann** [comment]
Please also log the value of these config options as discussed!

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 12]
```diff
@@ -471,7 +508,11 @@ void RocksDBOptionFeature::start() {
       << ", num_threads_high: " << _numThreadsHigh
       << ", num_threads_low: " << _numThreadsLow << ", block_cache_size: " << _blockCacheSize
       << ", block_cache_shard_bits: " << _blockCacheShardBits
-      << ", block_cache_strict_capacity_limit: " << _enforceBlockCacheSizeLimit
+      << ", ck_cache_strict_capacity_limit: " << _enforceBlockCacheSizeLimit
+      << ", cache_index_and_filter_blocks: " << _cacheIndexAndFilterBlocks
+      << ", cache_index_and_filter_blocks_with_high_priority: " << _cacheIndexNFilterBlocksWithHighPriority
+      << ", pin_l0_filter_and_index_blocks_in_cache: " << _pinTopLevelIndexAndFilter
+      << ", pin_top_level_index_and_filter: " << _pinTopLevelIndexAndFilter
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:515]:
```suggestion
      << ", cache_index_and_filter_blocks_with_high_priority: " << std::boolalpha << _cacheIndexAndFilterBlocksWithHighPriority
      << ", pin_l0_filter_and_index_blocks_in_cache: " << std::boolalpha << _pinl0FilterAndIndexBlocksInCache
      << ", pin_top_level_index_and_filter: " << std::boolalpha << _pinTopLevelIndexAndFilter
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 13]
```diff
@@ -471,7 +508,11 @@ void RocksDBOptionFeature::start() {
       << ", num_threads_high: " << _numThreadsHigh
       << ", num_threads_low: " << _numThreadsLow << ", block_cache_size: " << _blockCacheSize
       << ", block_cache_shard_bits: " << _blockCacheShardBits
-      << ", block_cache_strict_capacity_limit: " << _enforceBlockCacheSizeLimit
+      << ", ck_cache_strict_capacity_limit: " << _enforceBlockCacheSizeLimit
+      << ", cache_index_and_filter_blocks: " << _cacheIndexAndFilterBlocks
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp:512]:
```suggestion
      << ", block_cache_strict_capacity_limit: " << std::boolalpha << _enforceBlockCacheSizeLimit
      << ", cache_index_and_filter_blocks: " << std::boolalpha << _cacheIndexAndFilterBlocks
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10193/

File: CHANGELOG [hunk 14]
```diff
@@ -1,6 +1,17 @@
 devel
 -----
 
+* added multiple rocksdb configuration options to arangodb:
```
Comment (jsteemann) [inline CHANGELOG:4]:
```suggestion
* Added multiple RocksDB configuration options to arangod:
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10198/

**Pull request merged**